ethcore/light/src/on_demand/request.rs

```diff
@@ -26,17 +26,383 @@ use ethcore::receipt::Receipt;
+		}
+	}
+
+	/// Will succeed if all fields have been filled, will fail otherwise.
+	fn complete(self) -> Result<Self::Complete, net_request::NoSuchOutput> {
+		use ::request::CompleteRequest;
+
+		match self {
+			CheckedRequest::HeaderProof(_, req) => req.complete().map(CompleteRequest::HeaderProof),
+			CheckedRequest::HeaderByHash(_, req) => req.complete().map(CompleteRequest::Headers),
+			CheckedRequest::Receipts(_, req) => req.complete().map(CompleteRequest::Receipts),
+			CheckedRequest::Body(_, req) => req.complete().map(CompleteRequest::Body),
+			CheckedRequest::Account(_, req) => req.complete().map(CompleteRequest::Account),
+			CheckedRequest::Code(_, req) => req.complete().map(CompleteRequest::Code),
+			CheckedRequest::Execution(_, req) => req.complete().map(CompleteRequest::Execution),
+		}
+	}
+
+
+	fn adjust_refs<F>(&mut self, mapping: F) where F: FnMut(usize) -> usize {
+		match *self {
+			CheckedRequest::HeaderProof(_, ref mut req) => req.adjust_refs(mapping),
+			CheckedRequest::HeaderByHash(_, ref mut req) => req.adjust_refs(mapping),
+			CheckedRequest::Receipts(_, ref mut req) => req.adjust_refs(mapping),
+			CheckedRequest::Body(_, ref mut req) => req.adjust_refs(mapping),
+			CheckedRequest::Account(_, ref mut req) => req.adjust_refs(mapping),
+			CheckedRequest::Code(_, ref mut req) => req.adjust_refs(mapping),
+			CheckedRequest::Execution(_, ref mut req) => req.adjust_refs(mapping),
+		}
+	}
+}
+
+impl net_request::CheckedRequest for CheckedRequest {
+	type Extract = Response;
+	type Error = Error;
+	type Environment = Mutex<::cache::Cache>;
+
+	/// Check whether the response matches (beyond the type).
+	fn check_response(&self, cache: &Mutex<::cache::Cache>, response: &Self::Response) -> Result<Response, Error> {
+		use ::request::Response as NetResponse;
+
+		// check response against contained prover.
+		match (self, response) {
+			(&CheckedRequest::HeaderProof(ref prover, _), &NetResponse::HeaderProof(ref res)) =>
+				prover.check_response(cache, &res.proof).map(Response::HeaderProof),
+			(&CheckedRequest::HeaderByHash(ref prover, _), &NetResponse::Headers(ref res)) =>
+				prover.check_response(cache, &res.headers).map(Response::HeaderByHash),
+			(&CheckedRequest::Receipts(ref prover, _), &NetResponse::Receipts(ref res)) =>
+				prover.check_response(cache, &res.receipts).map(Response::Receipts),
+			(&CheckedRequest::Body(ref prover, _), &NetResponse::Body(ref res)) =>
+				prover.check_response(cache, &res.body).map(Response::Body),
+			(&CheckedRequest::Account(ref prover, _), &NetResponse::Account(ref res)) =>
+				prover.check_response(cache, &res.proof).map(Response::Account),
+			(&CheckedRequest::Code(ref prover, _), &NetResponse::Code(ref res)) =>
+				prover.check_response(cache, &res.code).map(Response::Code),
+			(&CheckedRequest::Execution(ref prover, _), &NetResponse::Execution(ref res)) =>
+				prover.check_response(cache, &res.items).map(Response::Execution),
+			_ => Err(Error::WrongKind),
```

Comment:
I think it would be more future proof to enumerate all `CheckedRequest` (so that when the protocol is extended you immediatelly see all places requiring implementation), but I know it would make the code here a bit more ugly (and macro would be required):

```rust
match *self {
  CheckedRequest::Account(..) => match!(NetResponse::Account(), response) {
  }
}
```